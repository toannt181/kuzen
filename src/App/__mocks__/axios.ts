import { AxiosResponse } from 'axios'
import { IUserInfo } from '../../types'

// load our test.json file. This can be copied to the local 
// folder. Can be a short version of your actual data set.
const data : IUserInfo = {
  login: 'hi',
  public_repos: 10,
}

// Our mocked response
const axiosResponse: AxiosResponse = {
  data,
  status: 200,
  statusText: 'OK',
  config: {},
  headers: {},
}

// axios mocked
export default {
  // Typescript requires a 'default'
  default: {
    get: jest.fn().mockImplementation(() => Promise.resolve(axiosResponse)),
  },
  defaults: {},
  get: jest.fn(() => Promise.resolve(axiosResponse)),
}
