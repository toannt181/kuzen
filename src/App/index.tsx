import React, { FC, useCallback, useState, useEffect, useRef } from 'react'
import RepositoryList from './containers/RepositoryList'
import * as Services from '../services'
import { IUserInfo, THOTTLE_TIME } from '../types'

type Props = {}

const RepositoryContainer: FC<Props> = () => {
  const [username, setUsername] = useState('')
  const [userInfo, setUserInfo] = useState<IUserInfo | null>(null)

  const requestId = useRef<NodeJS.Timeout>()

  const fetchUserInfo = async (username: string) => {
    try {
      const user = await Services.fetchUserInfo(username)
      setUserInfo(user.data)
    } catch {
      setUserInfo(null)
    }
  }

  useEffect(() => {
    if (!username.trim()) return

    requestId.current = setTimeout(() => fetchUserInfo(username), THOTTLE_TIME)

    return () => {
      if (requestId.current) {
        clearTimeout(requestId.current)
      }
    }
  }, [username])

  const onChangeUsername = useCallback((e) => {
    setUsername(e.target.value)
  }, [])

  return (
    <div className="container">
      <div className="field">
        <label className="label">Github username</label>
        <div className="control">
          <input className="input" type="text" aria-label="username-input" onChange={onChangeUsername} value={username} />
        </div>
      </div>

      {userInfo && <RepositoryList userInfo={userInfo} />}
    </div>
  )
}

export default RepositoryContainer
