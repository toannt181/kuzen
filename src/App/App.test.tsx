import React from 'react'
import { render, fireEvent, cleanup, act } from '@testing-library/react'
import App from '.'
import axios from 'axios'

jest.mock('axios')

describe(('On test App'), () => {
  afterEach(cleanup)

  test('should match snapshot', () => {
    const { asFragment } = render(<App />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('should display input changes when user typing and request api', async (done) => {
    const { getByLabelText } = render(<App />)
    const input = getByLabelText('username-input') as HTMLInputElement
    await act(async () => {
      fireEvent.change(input, { target: { value: 'hi' } })
    })
    expect(input.value).toBe('hi')
    setTimeout(() => {
      expect(axios.get).toHaveBeenCalledTimes(1)
      done()
    }, 1000)
  })
})
