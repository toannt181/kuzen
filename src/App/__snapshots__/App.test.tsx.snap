// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`On test App should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        Github username
      </label>
      <div
        class="control"
      >
        <input
          aria-label="username-input"
          class="input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="field"
    >
      <label
        class="label"
      >
        Github username
      </label>
      <div
        class="control"
      >
        <input
          aria-label="username-input"
          class="input"
          type="text"
          value=""
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
